arrow-syntax/src/main/kotlin/arrow/syntax/function/memoization.kt

```diff
@@ -1,117 +1,208 @@
 package arrow.syntax.function
 
-import arrow.syntax.internal.Platform.newConcurrentMap
-
+import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.loop
+
+/**
+ * Memoizes the given **pure** function so that invocations with the same arguments will only execute the function once.
+ *
+ * ```kotlin:ank:playground
+ * import arrow.syntax.function.memoize
+ * fun someWorkIntensiveFunction(someParam: Int): String = "$someParam"
+ *
+ * fun main() {
+ *   //sampleStart
+ *   val memoizedF = ::someWorkIntensiveFunction.memoize()
+ *
+ *   // The first invocation will store the argument and result in a cache inside the `memoizedF` reference.
+ *   val value1 = memoizedF(42)
+ *   // This second invocation won't really call the `someWorkIntensiveFunction` function
+ *   //but retrieve the result from the previous invocation instead.
+ *   val value2 = memoizedF(42)
+ *
+ *   //sampleEnd
+ *   println("$value1 $value2")
+ * }
+ * ```
+ *
+ * Note that calling this function with the same parameters in parallel might cause the function to be executed twice.
+ */
 fun <R> (() -> R).memoize(): () -> R = object : () -> R {
```

Comment:
Does it make sense to define them as suspended functions. Given the nature of memoizing, or am I mistaken here? 
cc @nomisRev

Reply:
This PR just updates the existing code to remove the dependency on Java classes. We can do a variant of this for suspending code as well.
Kinda silly that this needs to be defined separately tho, since normal code should work for suspending code. Only the other way around is not true.

Reply:
Actually, the `suspend` version is the only pure/referential transparent implementation since it involves allocation mutable shared state to do the memoisation.

Reply:
Yes, that was one of my concerns. 
Maybe it's worth deprecating it first and rolling out a suspended version either here or in `Arrow Fx`